addons/hr_payroll/models/hr_payslip.py

```diff
@@ -0,0 +1,545 @@
+                        'register_id': rule.register_id.id,
+                        'amount': amount,
+                        'employee_id': contract.employee_id.id,
+                        'quantity': qty,
+                        'rate': rate,
+                    }
+                else:
+                    #blacklist this rule and its children
+                    blacklist += [id for id, seq in rule._recursive_search_of_rules()]
+
+        result = [value for code, value in result_dict.items()]
+        return result
+
+    @api.multi
+    @api.onchange('employee_id', 'date_from')
+    def onchange_employee_id_wrapper(self):
+        self.ensure_one()
+        values = self.onchange_employee_id(self.date_from, self.date_to, self.employee_id.id, self.contract_id.id)['value']
+        for fname, value in values.iteritems():
+            setattr(self, fname, value)
+
+    @api.multi
+    def onchange_employee_id(self, date_from, date_to, employee_id=False, contract_id=False):
+        #delete old worked days lines
+        if self.worked_days_line_ids:
+            self.worked_days_line_ids.unlink()
+        #delete old input lines
+        if self.input_line_ids:
+            self.input_line_ids.unlink()
+
+        #defaults
+        res = {'value':{
+                      'line_ids':[],
+                      'input_line_ids': [],
+                      'worked_days_line_ids': [],
+                      #'details_by_salary_head':[], TODO put me back
+                      'name':'',
+                      'contract_id': False,
+                      'struct_id': False,
+                      }
+            }
+
+        if (not employee_id) or (not date_from) or (not date_to):
+            return res
+        ttyme = datetime.fromtimestamp(time.mktime(time.strptime(date_from, "%Y-%m-%d")))
+        employee_id = self.env['hr.employee'].browse(employee_id)
+        res['value'].update({
+                    'name': _('Salary Slip of %s for %s') % (employee_id.name, tools.ustr(ttyme.strftime('%B-%Y'))),
+                    'company_id': employee_id.company_id.id
+        })
+
+        if not self.env.context.get('contract', False):
+            #fill with the first contract of the employee
+            contracts = self.get_contract(employee_id, date_from, date_to)
+        else:
+            if contract_id:
+                #set the list of contract for which the input have to be filled
+                contracts = self.env['hr.contract'].browse(contract_id)
+            else:
+                #if we don't give the contract, then the input to fill should be for all current contracts of the employee
+                contracts = self.get_contract(employee_id, date_from, date_to)
+
+        if not contracts:
+            return res
+        contract = contracts[0]
+        res['value'].update({
+                    'contract_id': contract.id
+        })
+        struct_record = contract and contract.struct_id or False
```

Comment:
contract is recorset we can directly write like : 
 struct_record = contract.struct_id